src/routers/user-router.ts

```diff
@@ -0,0 +1,130 @@
+import { Request, Response, Router } from 'express';
+import { v4 as uuidv4 } from 'uuid';
+
+import asyncHandler from '../handle-midleware/utils';
+import { validateSchema } from '../validation/validation-middleware';
+
+import * as users from '../../data/data.json';
+import { userSchema } from '../validation/schema';
+
+export const router = Router();
+
+export interface User {
+    id: string;
+    login: string;
+    password: string;
+    age: number;
+    isDeleted: boolean;
+}
+
+const getAutoSuggestUsers = (loginSubstring: string, limit: string) =>
+    users.users
+        .filter((user: User) => !user.isDeleted)
+        .sort((user1: User, user2: User) => {
+            if (user1.login < user2.login) {
+                return -1;
+            }
+            if (user1.login > user2.login) {
+                return 1;
+            }
+            return 0;
+        })
+        .filter((user: User) => user.login.includes(loginSubstring))
+        .slice(0, +limit);
+
+
+router.get(
+    '/users',
+    asyncHandler(async (req: Request, res: Response) => {
+        if (users.users) {
+            res.json(users.users);
+        } else {
+            res.status(404);
+        }
+        res.end();
+    })
+);
+
+router.get(
+    '/users/:id',
+    asyncHandler(async (req: Request, res: Response) => {
+        const id = req.params.id;
+        const userById = users.users.find((item: User) => item.id === id);
+
+        if (userById) {
+            res.json(userById);
+        } else {
+            res.status(404);
+        }
+        res.end();
+    })
+);
+
+router.get(
+    '/suggestUsers',
+    asyncHandler(async (req: Request, res: Response) => {
+
+        const { loginSubstring, limit } = req.query as { [key: string]: string };
+        const listUsers = getAutoSuggestUsers(loginSubstring, limit);
+
+        if (listUsers.length) {
+            res.json(listUsers);
+        } else {
+            res.status(404);
+        }
+        res.end();
+    })
+);
+
+router.post(
+    '/users',
+    validateSchema(userSchema, users),
+    asyncHandler(async (req: Request, res: Response) => {
+        const content = req.body;
+        content.id = uuidv4();
+        users.users.push(content);
+
+        res.status(201);
+        res.end();
+    })
+);
+
+router.put(
+    '/users/:id',
+    validateSchema(userSchema, users),
+    asyncHandler(async (req: Request, res: Response) => {
+        const id = req.params.id;
+        const content = req.body;
+        const index = users.users.findIndex((item: User) => item.id === id);
```

Comment:
maybe better use `find` insted of `findIndex` if you want to mutate data

Reply:
fixed